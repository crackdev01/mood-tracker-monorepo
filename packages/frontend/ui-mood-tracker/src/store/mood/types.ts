// Define Mood Store State.
export interface MoodState {
  mood: [];
}

// Define Mood Store actions.
export enum MoodActions {
  FETCH_MOODS = 'FETCH_MOODS',
  RENDER_MOODS = 'RENDER_MOODS',
}
