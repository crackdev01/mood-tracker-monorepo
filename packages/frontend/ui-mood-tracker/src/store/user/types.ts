export interface UserRequest {
  username: string;
  password: string;
}

export interface Jwt {
  exp: number;
  iat: number;
  username: string;
  uuid: string;
}

export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface User {
  accessToken: string;
  decodedAccessToken: Jwt | null;
  // We set the coordinate values if user has provided permission.
  // Else we set to false.
  location?: Coordinates | string;
}

// Define User Store State.
export interface UserState {
  user: User;
}

// Define User Store actions.
export enum UserActions {
  FETCH_USER = 'FETCH_USER',
  RENDER_USER = 'RENDER_USER',
  LOGOUT_USER = 'LOGOUT_USER',
  SET_LOCATION_SUCCESS = 'SET_LOCATION_SUCCESS',
  SET_LOCATION_DENIED = 'SET_LOCATION_DENIED',
  SET_LOCATION_ERROR = 'SET_LOCATION_ERROR',
}
