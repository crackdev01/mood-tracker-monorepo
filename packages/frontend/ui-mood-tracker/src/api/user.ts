import { moodTrackerApi } from './core';

export async function authenticateUser(): Promise<any> {
  return await moodTrackerApi.post('/auth/login', { username: 'testuser', password: 'password' });
}
