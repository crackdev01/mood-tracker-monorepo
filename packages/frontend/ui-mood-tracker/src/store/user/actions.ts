import { UserRequest, UserActions } from './types';

// FIXME: This should not require any payload.
export const fetchUser = (payload: UserRequest) => {
  return {
    type: UserActions.FETCH_USER,
    payload,
  };
};

export const renderUser = (payload: any) => {
  return {
    type: UserActions.RENDER_USER,
    payload,
  };
};
