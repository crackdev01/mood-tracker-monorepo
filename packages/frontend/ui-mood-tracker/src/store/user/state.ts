import { UserState } from './types';

export const userState: UserState = {
  user: {},
};
