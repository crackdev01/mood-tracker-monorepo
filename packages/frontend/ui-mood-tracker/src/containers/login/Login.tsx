import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch } from 'react-redux';
import { Button, Header, Form, Input } from 'semantic-ui-react';

import './login.scss';
import { UserActions } from '../../store/user/types';

const Login = () => {
  const dispatch = useDispatch();
  const { t } = useTranslation(['Login']);
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const loading = false;

  const updateUsername = (event: any, data: any) => {
    setUsername(data.value);
  };

  const updatePassword = (_: any, data: any) => {
    setPassword(data.value);
  };

  const login = () => {
    const payload = { username, password };
    dispatch({
      type: UserActions.FETCH_USER,
      payload,
    });
  };

  return (
    <article className="login">
      <Form className="login__form">
        <Header as="h2" className="login__form__header">
          {t('header')}
        </Header>

        <Form.Field className="login__form__input">
          <label>{t('username')}</label>
          <Input id="username" placeholder={t('username')} onChange={updateUsername} />
        </Form.Field>

        <Form.Field className="login__form__input">
          <label>{t('password')}</label>
          <Input id="password" placeholder={t('password')} onChange={updatePassword} />
        </Form.Field>

        <Button className="login__form__input" loading={loading} onClick={login}>
          {t('submit')}
        </Button>
      </Form>
    </article>
  );
};

export default Login;
