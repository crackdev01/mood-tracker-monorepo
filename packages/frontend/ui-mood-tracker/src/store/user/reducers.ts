import { userState } from './state';
import { UserActions } from './types';

export const user = (state = userState, action: any) => {
  switch (action.type) {
    case UserActions.FETCH_USER:
      return {
        ...state,
      };
    case UserActions.RENDER_USER:
      return {
        ...state,
        user: action.data,
      };
    default:
      return state;
  }
};
