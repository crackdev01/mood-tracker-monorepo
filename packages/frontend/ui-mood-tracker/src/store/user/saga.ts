import { all, call, put, takeEvery, fork } from 'redux-saga/effects';
import jwt_decode from 'jwt-decode';

import { User, UserActions } from './types';
import { userState } from './state';
import { authenticateUser } from '../../api/user';

export function* loginUserCall(action: any) {
  const { data } = yield call(authenticateUser, action.payload);
  const user: User = {
    accessToken: data.accessToken,
    decodedAccessToken: jwt_decode(data.accessToken),
  };
  yield put({ type: UserActions.RENDER_USER, data: user });
}

export function* loginUser() {
  yield takeEvery(UserActions.FETCH_USER, loginUserCall);
}

export function* setUserLocationSuccessCall(action: any) {
  const { location } = action.payload;
  yield put({ type: UserActions.RENDER_USER, data: location });
}

export function* setUserLocationSuccess() {
  yield takeEvery(UserActions.SET_LOCATION_SUCCESS, setUserLocationSuccessCall);
}

export function* setUserLocationErrorCall() {
  const location = UserActions.SET_LOCATION_ERROR;
  yield put({ type: UserActions.RENDER_USER, data: location });
}

export function* setUserLocationError() {
  yield takeEvery(UserActions.SET_LOCATION_ERROR, setUserLocationErrorCall);
}

export function* setUserLocationDeniedCall() {
  const location = UserActions.SET_LOCATION_DENIED;
  yield put({ type: UserActions.RENDER_USER, data: location });
}

export function* setUserLocationDenied() {
  yield takeEvery(UserActions.SET_LOCATION_DENIED, setUserLocationDeniedCall);
}

export function* logoutUserCall() {
  // Reset use store to default state to trigger logout.
  yield put({ type: UserActions.RENDER_USER, data: userState });
}

export function* logoutUser() {
  yield takeEvery(UserActions.LOGOUT_USER, logoutUserCall);
}

export function* userSaga() {
  yield all([fork(loginUser), fork(logoutUser)]);
}

export default userSaga;
