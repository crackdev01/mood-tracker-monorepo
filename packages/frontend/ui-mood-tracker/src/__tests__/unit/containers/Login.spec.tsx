import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Login from '../../../containers/login/Login';
import { UserActions } from '../../../store/user/types';
import { mockUserPayload } from '../../fixtures';

Enzyme.configure({ adapter: new Adapter() });

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch,
}));

describe('Login', () => {
  test('renders', () => {
    const w = shallow(<Login />);
    expect(w.debug()).toMatchSnapshot();
  });

  test('triggers login', () => {
    mockDispatch.mockReturnValueOnce({});

    const w = shallow(<Login />);
    w.find('Input').at(0).simulate('change', undefined, { value: mockUserPayload.username });
    w.find('Input').at(1).simulate('change', undefined, { value: mockUserPayload.password });
    w.find('Button').at(0).simulate('click');

    expect(mockDispatch).toHaveBeenCalledWith({
      type: UserActions.FETCH_USER,
      payload: mockUserPayload,
    });
  });
});
