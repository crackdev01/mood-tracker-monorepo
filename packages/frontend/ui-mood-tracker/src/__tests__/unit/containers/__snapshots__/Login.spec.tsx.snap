// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders 1`] = `
"<article className=\\"login\\">
  <Form className=\\"login__form\\" as=\\"form\\">
    <Header as=\\"h2\\" className=\\"login__form__header\\">
      header
    </Header>
    <FormField className=\\"login__form__input\\">
      <label>
        username
      </label>
      <Input placeholder=\\"username\\" onChange={[Function: updateUsername]} type=\\"text\\" />
    </FormField>
    <FormField className=\\"login__form__input\\">
      <label>
        password
      </label>
      <Input placeholder=\\"password\\" onChange={[Function: updatePassword]} type=\\"text\\" />
    </FormField>
    <Button className=\\"login__form__input\\" loading={false} onClick={[Function: login]} as=\\"button\\">
      submit
    </Button>
  </Form>
</article>"
`;
